### PR TITLE
Supports non-Canvas rendering.

### PR DESCRIPTION
**Pull request by MikuAuahDark**
Added `baseAtlas:useImageData(boolean)` to tell if this atlas should use ImageData rendering or not. ImageData rendering is default if `love.graphics` is not loaded. That function can only be called if the atlas contents are empty; you can't switch between ImageData and Canvas rendering if there's at least one texture in the atlas.

~~I haven't tested this (will edit when I did).~~ Please let me know it's ready to merge or if there some changes I need to apply.

### CODE REVIEW

File: fixedSize.lua [hunk 1]
```diff
@@ -30,56 +32,78 @@ fixedSizeTA.bake = function(self)
...
+      data = newImageData(widthCanvas, heightCanvas, "rgba8")
+      for x=0, columns-1 do
+        for y=0, rows-1 do
+          local index = (x+y*rows)+1
+          if index > maxIndex then
+            break
+          end
+          local x, y = x * widthPadded + self.padding + self.extrude, y * heightPadded + self.padding + self.extrude
+          local image = self.images[index]
+          data:paste(image, x, y, 0, 0, image:getDimensions())
```
Comment (EngineerSmith) [inline fixedSize.lua]: image is a table containing the image, it should be `image.image` and `image.image:getDimensions()`

File: fixedSize.lua [hunk 2]
```diff
@@ -30,56 +32,78 @@ fixedSizeTA.bake = function(self)
...
+          if self.extrude > 0 then
+            util.extrudeWithFill(data, image, self.extrude, x, y)
```
Comment (EngineerSmith) [inline fixedSize.lua]: Function expects imageData, should be `image.image`

File: util.lua [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  end
+end
+
+util.extrudeWithFill = function(dest, src, n, x, y)
+  local iw, ih = src:getDimensions()
+  extrudeImageData(dest, src, n, x, y, 0, -1, 0, 0, iw, 1) -- top
+  extrudeImageData(dest, src, n, x, y, -1, 0, 0, 0, 1, ih) -- left
+  extrudeImageData(dest, src, n, x, y + ih - 1, 0, 1, 0, ih - 1, iw, 1) -- bottom
+  extrudeImageData(dest, src, n, x + iw - 1, y, 1, 0, iw - 1, 0, 1, ih) -- right
+  fillImageData(dest, x - n - 1, y - n - 1, n, n, src:getPixel(0, 0)) -- top-left
```
Comment (EngineerSmith) [inline util.lua]: Top-left, Top-right and Bottom-left are misalligned for both fixedSized and dynamicSized
1 extrude image
2 extrude image

**Comment by EngineerSmith** [comment]
I've made a headless [test repo](https://github.com/EngineerSmith/TA-tests) - you can drop a version of RTA within and it'll output atlases in the save directory

**Comment by MikuAuahDark** [comment]
I've fixed the issue and tested it by running the Canvas rendering once and use it as the reference. I store the Canvas result in `correct` folder and the result is they're 100% identical.

Here's the `magick compare` output:
<details>
<summary>Large output log</summary>

```
C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>for %a in (*.png) do magick compare -verbose -metric mae correct/%a %
a png:nul

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/dynamicSize_0_0_0.png dynamicSize_0_0_0.png png:nul
correct/dynamicSize_0_0_0.png PNG 48x16 48x16+0+0 8-bit sRGB 2041B 0.000u 0:00.000
dynamicSize_0_0_0.png PNG 48x16 48x16+0+0 8-bit sRGB 2041B 0.000u 0:00.000
Image: correct/dynamicSize_0_0_0.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    all: 0 (0)
correct/dynamicSize_0_0_0.png=>nul PNG 48x16 8-bit sRGB 2041B 0.000u 0:00.007

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/dynamicSize_0_0_1.png dynamicSize_0_0_1.png png:nul
correct/dynamicSize_0_0_1.png PNG 50x16 50x16+0+0 8-bit sRGB 2263B 0.000u 0:00.001
dynamicSize_0_0_1.png PNG 50x16 50x16+0+0 8-bit sRGB 2263B 0.000u 0:00.000
Image: correct/dynamicSize_0_0_1.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/dynamicSize_0_0_1.png=>nul PNG 50x16 8-bit sRGB 2263B 0.000u 0:00.004

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/dynamicSize_0_1_0.png dynamicSize_0_1_0.png png:nul
correct/dynamicSize_0_1_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2109B 0.000u 0:00.000
dynamicSize_0_1_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2109B 0.000u 0:00.000
Image: correct/dynamicSize_0_1_0.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    all: 0 (0)
correct/dynamicSize_0_1_0.png=>nul PNG 54x18 8-bit sRGB 2109B 0.000u 0:00.004

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/dynamicSize_1_0_0.png dynamicSize_1_0_0.png png:nul
correct/dynamicSize_1_0_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2314B 0.000u 0:00.000
dynamicSize_1_0_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2314B 0.000u 0:00.000
Image: correct/dynamicSize_1_0_0.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/dynamicSize_1_0_0.png=>nul PNG 54x18 8-bit sRGB 2314B 0.000u 0:00.006

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/dynamicSize_1_1_1.png dynamicSize_1_1_1.png png:nul
correct/dynamicSize_1_1_1.png PNG 62x20 62x20+0+0 8-bit sRGB 2377B 0.000u 0:00.000
dynamicSize_1_1_1.png PNG 62x20 62x20+0+0 8-bit sRGB 2377B 0.000u 0:00.000
Image: correct/dynamicSize_1_1_1.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/dynamicSize_1_1_1.png=>nul PNG 62x20 8-bit sRGB 2377B 0.000u 0:00.005

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/dynamicSize_5_2_5.png dynamicSize_5_2_5.png png:nul
correct/dynamicSize_5_2_5.png PNG 100x30 100x30+0+0 8-bit sRGB 2472B 0.016u 0:00.001
dynamicSize_5_2_5.png PNG 100x30 100x30+0+0 8-bit sRGB 2472B 0.000u 0:00.000
Image: correct/dynamicSize_5_2_5.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/dynamicSize_5_2_5.png=>nul PNG 100x30 8-bit sRGB 2472B 0.000u 0:00.006

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/fixedSize_0_0_0.png fixedSize_0_0_0.png png:nul
correct/fixedSize_0_0_0.png PNG 48x16 48x16+0+0 8-bit sRGB 2041B 0.000u 0:00.000
fixedSize_0_0_0.png PNG 48x16 48x16+0+0 8-bit sRGB 2041B 0.000u 0:00.000
Image: correct/fixedSize_0_0_0.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    all: 0 (0)
correct/fixedSize_0_0_0.png=>nul PNG 48x16 8-bit sRGB 2041B 0.000u 0:00.004

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/fixedSize_0_0_1.png fixedSize_0_0_1.png png:nul
correct/fixedSize_0_0_1.png PNG 50x16 50x16+0+0 8-bit sRGB 2263B 0.000u 0:00.000
fixedSize_0_0_1.png PNG 50x16 50x16+0+0 8-bit sRGB 2263B 0.000u 0:00.000
Image: correct/fixedSize_0_0_1.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/fixedSize_0_0_1.png=>nul PNG 50x16 8-bit sRGB 2263B 0.016u 0:00.005

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/fixedSize_0_1_0.png fixedSize_0_1_0.png png:nul
correct/fixedSize_0_1_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2109B 0.000u 0:00.000
fixedSize_0_1_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2109B 0.000u 0:00.000
Image: correct/fixedSize_0_1_0.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    all: 0 (0)
correct/fixedSize_0_1_0.png=>nul PNG 54x18 8-bit sRGB 2109B 0.000u 0:00.005

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/fixedSize_1_0_0.png fixedSize_1_0_0.png png:nul
correct/fixedSize_1_0_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2314B 0.000u 0:00.000
fixedSize_1_0_0.png PNG 54x18 54x18+0+0 8-bit sRGB 2314B 0.000u 0:00.000
Image: correct/fixedSize_1_0_0.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/fixedSize_1_0_0.png=>nul PNG 54x18 8-bit sRGB 2314B 0.000u 0:00.005

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/fixedSize_1_1_1.png fixedSize_1_1_1.png png:nul
correct/fixedSize_1_1_1.png PNG 62x20 62x20+0+0 8-bit sRGB 2377B 0.000u 0:00.000
fixedSize_1_1_1.png PNG 62x20 62x20+0+0 8-bit sRGB 2377B 0.000u 0:00.000
Image: correct/fixedSize_1_1_1.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/fixedSize_1_1_1.png=>nul PNG 62x20 8-bit sRGB 2377B 0.000u 0:00.005

C:\Users\MikuAuahDark\AppData\Roaming\LOVE\TA TEST>magick compare -verbose -metric mae correct/fixedSize_5_2_5.png fixedSize_5_2_5.png png:nul
correct/fixedSize_5_2_5.png PNG 100x30 100x30+0+0 8-bit sRGB 2472B 0.000u 0:00.000
fixedSize_5_2_5.png PNG 100x30 100x30+0+0 8-bit sRGB 2472B 0.000u 0:00.000
Image: correct/fixedSize_5_2_5.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
correct/fixedSize_5_2_5.png=>nul PNG 100x30 8-bit sRGB 2472B 0.000u 0:00.005
```
</details>

**Pull request merged**